src/SimpleSMT/Solver.hs

```diff
@@ -0,0 +1,395 @@
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | A module for interacting with an SMT solver, using SmtLib-2 format.
+module SimpleSMT.Solver
+    -- * Basic Solver Interface
+  ( Solver(..)
+  , Backend(..)
+  , initSolverWith
+  , command
+  , ackCommand
+  , simpleCommand
+  , simpleCommandMaybe
+  , loadFile
+    -- * Common SmtLib-2 Commands
+  , setLogic
+  , setLogicMaybe
+  , setOption
+  , setOptionMaybe
+  , produceUnsatCores
+  , push
+  , pushMany
+  , pop
+  , popMany
+  , inNewScope
+  , declare
+  , declareFun
+  , declareDatatype
+  , define
+  , defineFun
+  , defineFunRec
+  , defineFunsRec
+  , assert
+  , check
+  , getExprs
+  , getExpr
+  , getConsts
+  , getConst
+  , getUnsatCore
+  ) where
+
+import SimpleSMT.SExpr
+import Prelude hiding (not, and, or, abs, div, mod, concat, const, log)
+import qualified Control.Exception as X
+import qualified Data.ByteString.Lazy.Char8 as LBS
+import Data.ByteString.Builder (Builder)
+import Data.IORef (IORef, newIORef, readIORef, writeIORef, atomicModifyIORef)
+
+data Backend = Backend {
+  send :: LBS.ByteString -> IO ()
+  -- ^ Send a command to the backend.
+  , response :: IORef LBS.ByteString
+  -- ^ A buffer holding the solver's responses to the commands.
+  }
+
+type Queue = IORef Builder
+
+-- | Push a command on the solver's queue of commands to evaluate.
+-- The command must not produce any output when evaluated, unless it is the last
+-- command added before the queue is flushed.
+putQueue :: Queue -> SExpr -> IO ()
+putQueue q expr = atomicModifyIORef q $ \cmds ->
+  (cmds <> renderSExpr expr, ())
+
+-- | Empty the queue of commands to evaluate and return its content as a bytestring
+-- builder.
+flushQueue :: Queue -> IO Builder
+flushQueue q = atomicModifyIORef q $ \cmds ->
+    (mempty, cmds)
+
+data Solver = Solver
+  { backend :: Backend
+  -- ^ The backend processing the commands.
+  , queue :: Maybe Queue
+  -- ^ A queue to write commands that are to be sent to the solver lazily.
+  , log :: LBS.ByteString -> IO ()
+  -- ^ The function used for logging the solver's activity.
+  }
+
+-- | Send a command to the solver.
+sendSolver :: Solver -> LBS.ByteString -> IO ()
```

Comment:
Does the function return only after the solver produces a response? Otherwise, how do we know when `recvSolver` is good to call?

Reply:
I'm wondering what is the motivation to have `sendSolver` and `recvSolver`. I would have expected something like:
```
sendSolver :: Solver -> LBS.ByteString -> IO Sexp

-- | Like 'sendSolver_' but discards the result.
sendSolver_ :: Solver -> LBS.ByteString -> IO ()
```

Reply:
> Does the function return only after the solver produces a response?

This is true for the `Z3` backend and I expect it to be true for the `Process` backend as well because [this is how simple-smt works internally](https://github.com/yav/simple-smt/blob/dd03ca119ddcd50bde52fbe40c96eb60cc1d6338/SimpleSMT.hs#L319-L321) (as far as I understand). I agree that this is not ideal, and I had planned to say a word about this in the documentation.

Reply:
As to why I did this, here's my reasoning. Before making this change the interface provided a single
```
send :: Solver -> LBS.ByteString -> IO SExpr
```
I'd rather the output type be `LBS.ByteString` as well, so that parsing and logging the output can be taken care of in the common interface. 

The problem with that is that it's not really compatible with the `Process` backend. Since you cannot assume that the solver writes its output on a single-line (Z3 does not and I didn't find a way to configure it to do so), your only option to read the output is to use `hGetContents` and parse it until you get a full s-expression. So if you wanted to return an `LBS.ByteString` you'd have to parse the bytestring as an s-expression, serialize it back to a bytestring, only for it to be parsed again in the common interface.

My solution to that is to use this `response :: IORef ByteString` buffer, and the `sendSolver` and `recvSolver` functions that are deduced from it. To the user this shouldn't change anything because `command` is just `sendSolver >> recvSolver`, and this also has the advantage that parsing the `(success)`s when sending batch of commands is also implementable.

Reply:
A solution to have
```haskell
send :: Solver -> LBS.ByteString -> IO LBS.ByteString
```
would be to reimplement [this](https://hackage.haskell.org/package/bytestring-0.11.3.1/docs/src/Data.ByteString.html#hGetContentsSizeHint) for the `Process` backend.

Reply:
Oh I just realized was probably just about what functions are available in the API ? `sendSolver` and `recvSolver` aren't currently exposed in the API anyways, I just separated them because I thought it was cleaner but it's not particularly necessary.

Reply:
Your explanation makes sense to me. I think, though, that maybe it is possible to deal with the output of z3 as discussed in #2.